### PR TITLE
provide a `ListEntries` method

### PR DESCRIPTION
**Pull request by schomatis**
```
The `coreunix.Adder` was calling the `ListNames` method and then converting the
list to `FSNode`s with `Child`. The `Child` method generates a cache entry for
each child retrieved, to avoid reduce memory use the `Adder` called the exported
`Uncache` method (added specifically for that case) which gave outside access to
the internal cache. For more details see: https://github.com/ipfs/go-
ipfs/pull/3027.

To prevent access to the internal cache a `ListEntries` method is provided with
the same logic extracted from the `coreunix.Adder`. This allows now to make the
`Uncache` method private, which will permit a cache refactor in the future
without breaking the API.
```

### CODE REVIEW

**Comment by schomatis** [comment]
@Stebalien As discussed in https://github.com/ipfs/go-ipfs/pull/3027 this seems like the least invasive way to decouple the cache from the user API and allow a future refactor.

File: dir.go [hunk 1]
```diff
@@ -280,6 +280,30 @@ func (d *Directory) ListNames(ctx context.Context) ([]string, error) {
...
+func (d *Directory) ListEntries(ctx context.Context) ([]FSNode, error) {
+	var entries []FSNode
+
+	names, err := d.ListNames(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, name := range names {
+		child, err := d.Child(name)
```
Comment (Stebalien) [inline dir.go]: This isn't going to be atomic.

File: dir.go [hunk 2]
```diff
@@ -280,6 +280,30 @@ func (d *Directory) ListNames(ctx context.Context) ([]string, error) {
...
+		if err != nil {
+			return nil, err
+		}
+		d.uncache(name)
+		// Remove the cache entry generated by `Child` to avoid using
+		// too much memory.
+		// TODO: Provide a `Child`-like method that doesn't cache the retrieved child.
```
Comment (Stebalien) [inline dir.go]: Again, this is more than *caching*. It records which files are "open". If an `FSNode` returned from multiple calls to `ListEntries` are used to to modify a file concurrently, we'll end up with two open file descriptors for writing (and each will clobber the other).

**Comment by schomatis** [comment]
> I really don't see how this fixes anything. Really, it just hides the call to `Uncache` so a call to `d.Child("something")` followed by a call to `d.ListEntries(ctx)` will secretly "uncache" the child.

That was *exactly* what I was aiming at, so as you say, I'm missing something then. I want to hide the implementation, this cache (or register or map) is not part of the API:

> Again, this is more than _caching_. It records which files are "open".

Could you expand on that please? Both `Directory.AddChild` and `File.Open` do not record the operation in the cache, could you point me to the code flow you're referring to?

> If an `FSNode` returned from multiple calls to `ListEntries` are used to to modify a file concurrently, we'll end up with two open file descriptors for writing (and each will clobber the other).

Yes, we can add a lock around `ListEntries`, but do note that the problem you're referring to is independent of this PR, we do not lock directories (already discussed in https://github.com/ipfs/go-mfs/issues/3#issuecomment-438089984), so *any* concurrent operations on a directory, such as a simple `Directory.Child`, already suffers for that same problem, that is not the responsibility of this PR to fix.

**Comment by Stebalien** [comment]
> File.Open

In this case, it doesn't need to. Each open FileDescriptor takes the lock, only releasing it on close. However, we do add the `File` (not FileDescriptor) to the directories "cache" when we open it.

> Directory.AddChild

This doesn't return an opened handle to a file or a directory.

---

Again, the issue here is that if someone calls `d.Child("a"), d.Uncache("a"), d.Child("a")`, they'll get two independent copies of "a" and will be able to open both for writing. Without that `Uncache` call, they'd get one and opening one for writing would prevent us from opening the other.

---

> Yes, we can add a lock around ListEntries, but do note that the problem you're referring to is independent of this PR, we do not lock directories (already discussed in #3 (comment)), so any concurrent operations on a directory, such as a simple Directory.Child, already suffers for that same problem, that is not the responsibility of this PR to fix.

The point I was trying to make in https://github.com/ipfs/go-ipfs/pull/3027#issuecomment-448836212 was that `ListEntries` *would* be useful if and only if it provided a snapshot/atomic guarantee. Otherwise, it's no better (well, not much better) than the current ListNames + Child API. Any API that returns an `FSNode` needs to register that `FSNode` somewhere to avoid having multiple `FSNode` instances for a single file/directory.

**Comment by schomatis** [comment]
> Again, the issue here is that if someone calls `d.Child("a"), d.Uncache("a"), d.Child("a")`, they'll get two independent copies of "a" and will be able to open both for writing. Without that `Uncache` call, they'd get one and opening one for writing would prevent us from opening the other.

So, I'm sure I'm missing something here but that just there looks like the *best* argument for making `Uncache` private and removing it from the MFS API, which is the sole intention behind this PR, `ListEntries` is just an excuse to accomplish that.

> The point I was trying to make in [ipfs/go-ipfs#3027 (comment)](https://github.com/ipfs/go-ipfs/pull/3027#issuecomment-448836212) was that `ListEntries` _would_ be useful if and only if it provided a snapshot/atomic guarantee. Otherwise, it's no better (well, not much better) than the current ListNames + Child API.

I don't mind at all adding the lock to the entire function if it helps move this forward, I do feel that preventing the user from accessing the cache is a considerable improvement, a `Child(); ListEntries()` call still does suffer from the cache inconsistency you mention, but it seems like it would be less hazardous than the `d.Child("a"), d.Uncache("a"), d.Child("a")` case we're allowing now, especially since `ListEntries` is used in only one place as an ad hoc solution to lower memory consumption in https://github.com/ipfs/go-ipfs/pull/3027, where I think we should actually be closing the MFS root and removing its reference altogether instead of mangling with its internal state.

---------------------------------------------

If you still think this isn't much of an improvement let's go ahead and close this PR to focus on other parts of MFS we could work on.

**Comment by Stebalien** [comment]
> So, I'm sure I'm missing something here but that just there looks like the best argument for making Uncache private and removing it from the MFS API, which is the sole intention behind this PR, ListEntries is just an excuse to accomplish that.

`Uncache` should *definitely* not be a part of the API. It only exists there today as a hack to work around the memory issue in add.

> If you still think this isn't much of an improvement let's go ahead and close this PR to focus on other parts of MFS we could work on.

At the moment, I consider this patch to be strictly worse. As-is, the user has to explicitly call `Uncache` (which should probably be called `UnsafeUncache` and be well documented). After this patch, the user can unknowingly call `Uncache` by calling `ListEntries`. That doesn't make it any safer, it just makes it less visible.

The simple fix is documentation: we need to make it clear that `Uncache` is totally unsafe and that users should be careful to only call it on files they know aren't open elsewhere. The next step is to find a good way to avoid it (e.g., by asking users to close File/Directory objects and refcounting).

**Comment by schomatis** [comment]
> The simple fix is documentation

Fair enough.

**Pull request closed without merging**